src/server/server_state.cc

```diff
@@ -208,4 +227,46 @@ ServerState* ServerState::SafeTLocal() {
 bool ServerState::ShouldLogSlowCmd(unsigned latency_usec) const {
   return slow_log_shard_.IsEnabled() && latency_usec >= log_slower_than_usec;
 }
+
+void ServerState::ConnectionsWatcherFb(util::ListenerInterface* main) {
+  while (true) {
+    util::fb2::NoOpLock noop;
+    if (watcher_cv_.wait_for(noop, 1s, [this] { return gstate_ == GlobalState::SHUTTING_DOWN; })) {
```

Comment:
running this every second can be heavy if server has many connections

Reply:
agree. that's why I added a TODO:
https://github.com/dragonflydb/dragonfly/pull/4407/files#diff-6032993b813fa531ddc738b01203e65ded0d630c65a5de559a6772b98fac5951R258

in fact, newest version of helio already allows partial traversal. I will fix the TODO in this PR.

Reply:
fixed